### PR TITLE
Update desktop projects to target .NET 6

### PR DESCRIPTION
**Pull request by peppy**
- [x] Depends on ppy/osu-framework#5024.

### CODE REVIEW

**Review by smoogipoo — DISMISSED** [review]

**Comment by bdach** [comment]
Aside from b581ca1 there was one more local implementation of a border colour transform in `MuteButton` that seemingly evaded test coverage but would crash at runtime; I've fixed that one in 4bd58cf.

After the above the only remaining game-side usage of `SRGBColour` is [here](https://github.com/ppy/osu/blob/342452dabfe208c6cbfbeabedaf4a61b498d6454/osu.Game/Overlays/Notifications/Notification.cs#L237) and looks safe.

Given that the border colour fade transform has been implemented locally twice then it should probably become a framework-level helper, but that can happen as a follow up I think.

**Comment by bdach** [comment]
I'm just going to note that merging this as-is may prevent some users who may still have unmigrated EF/sqlite databases from migrating to realm, as migration fails due to the following:

```
022-02-14 22:50:20 [error]: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (The type initializer for 'Microsoft.EntityFrameworkCore.Query.ResultOperators.Internal.TrackingExpressionNode' threw an exception.)
2022-02-14 22:50:20 [error]: ---> System.TypeInitializationException: The type initializer for 'Microsoft.EntityFrameworkCore.Query.ResultOperators.Internal.TrackingExpressionNode' threw an exception.
2022-02-14 22:50:20 [error]: ---> System.TypeInitializationException: The type initializer for 'Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions' threw an exception.
2022-02-14 22:50:20 [error]: ---> System.InvalidOperationException: Sequence contains more than one matching element
2022-02-14 22:50:20 [error]: at System.Linq.ThrowHelper.ThrowMoreThanOneMatchException()
2022-02-14 22:50:20 [error]: at System.Linq.Enumerable.TryGetSingle[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)
2022-02-14 22:50:20 [error]: at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
2022-02-14 22:50:20 [error]: at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.GetMethod(String name, Int32 parameterCount, Func`2 predicate)
2022-02-14 22:50:20 [error]: at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions..cctor()
```

(i.e. #15544).

Given that EF Core is only used for reading old DBs, maybe it is possible to bump it for the time being to unblock migrations? Not sure, haven't tested much.

Other than that, I've tested 3 OS's that I can (linux, windows, android) and on a quick look all look to be behaving well after the bump.

**Comment by frenzibyte** [comment]
I've update three more `ChangeFocus` usages:
 - `MetadataSection`, due to being performed at `LoadComplete` (target not alive).
 - `LoginPanel`, due to being performed after reconstructing hierarchy (target not alive).
 - `LoginOverlay`, due to being performed while the overlay itself is not present, see second part of https://github.com/ppy/osu-framework/issues/5023#issuecomment-1035625923.

**Comment by smoogipoo** [comment]
In regards to the EF issue, we should be able to upgrade the package since we’re not doing any complex operations with it anymore.

**Comment by peppy** [comment]
Will need to test EF migration on iOS with package upgrades. I'll look at that today.

**Comment by peppy** [comment]
If anyone is able to test EF migration on android, that would be exceedingly helpful.

Attached is a small database you can drop in your data folder if you don't have one any more. Make sure to also move away `client.realm` to allow the migration to actually run.

[client.db.zip](https://github.com/ppy/osu/files/8066397/client.db.zip)

**Comment by peppy** [comment]
I've tested iOS migration from EF to work as expected. The iOS test projects don't seem to build, but from what I can tell this may have already been broken. I'd need to rollback to the non-preview version of visual studio to test this, but I'm not sure it's important to getting this merged.

@frenzibyte maybe you can try building any of the test projects from the iOS slnf (or the main sln) and see if you have any luck?

Apart from that, this should be in a good state to go.

**Comment by PercyDan54** [comment]
> If anyone is able to test EF migration on android (after the recent commits), that would be exceedingly helpful.
> 
> Attached is a small database you can drop in your data folder if you don't have one any more. Make sure to also move away `client.realm` to allow the migration to actually run.
> 
> [client.db.zip](https://github.com/ppy/osu/files/8066397/client.db.zip)

It crashes even without mirgrating, idk if it's a problem with my setup
image

Run osu.Game.Tests also crashed, but osu.Desktop is fine
image

**Comment by peppy** [comment]
Please stop posting screenshots of silly IDEs if you're going to try and help. Text is what we need (use Ctrl+C to copy to clipboard, then paste in the browser).

**Comment by peppy** [comment]
Your IDE shows a detached hash that doesn't exist on this branch. Please check your git, it's probably in a bad state.

**Comment by PercyDan54** [comment]
Ok here is logs:
```
2022-02-15 08:24:15 [error]: Realm startup failed with unrecoverable error; starting with a fresh database. A backup of your database has been made.
2022-02-15 08:24:15 [error]: System.ArgumentException: The class ScoreInfo is not in the limited set of classes for this realm (Parameter 'T')
2022-02-15 08:24:15 [error]: at Realms.Helpers.Argument.Ensure(Boolean condition, String message, String paramName)
2022-02-15 08:24:15 [error]: at Realms.Realm.All[T]()
2022-02-15 08:24:15 [error]: at osu.Game.Database.RealmAccess.cleanupPendingDeletions() in D:\Dev\OSU\osu\osu.Game\Database\RealmAccess.cs:line 166
2022-02-15 08:24:15 [error]: at osu.Game.Database.RealmAccess..ctor(Storage storage, String filename, IDatabaseContextFactory efContextFactory) in D:\Dev\OSU\osu\osu.Game\Database\RealmAccess.cs:line 148
2022-02-15 08:27:07 [error]: An unhandled error has occurred.
2022-02-15 08:27:07 [error]: System.ArgumentException: The class ScoreInfo is not in the limited set of classes for this realm (Parameter 'T')
2022-02-15 08:27:07 [error]: at Realms.Helpers.Argument.Ensure(Boolean condition, String message, String paramName)
2022-02-15 08:27:07 [error]: at Realms.Realm.All[T]()
2022-02-15 08:27:07 [error]: at osu.Game.Database.RealmAccess.cleanupPendingDeletions() in D:\Dev\OSU\osu\osu.Game\Database\RealmAccess.cs:line 166
2022-02-15 08:27:07 [error]: at osu.Game.Database.RealmAccess..ctor(Storage storage, String filename, IDatabaseContextFactory efContextFactory) in D:\Dev\OSU\osu\osu.Game\Database\RealmAccess.cs:line 157
2022-02-15 08:27:07 [error]: at osu.Game.OsuGameBase.load(ReadableKeyCombinationProvider keyCombinationProvider) in D:\Dev\OSU\osu\osu.Game\OsuGameBase.cs:line 201
2022-02-15 08:27:07 [error]: --- End of stack trace from previous location ---
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.BackgroundDependencyLoaderAttribute.<>c__DisplayClass6_0.<CreateActivator>b__4(Object target, IReadOnlyDependencyContainer dc)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.Activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyContainer.Inject[T](T instance)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Drawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Drawable.load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Drawable.Load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies, Boolean isDirectAsyncContext)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.loadChild(Drawable child)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.load(ShaderManager shaders, Nullable`1 cancellation)
2022-02-15 08:27:07 [error]: --- End of stack trace from previous location ---
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.BackgroundDependencyLoaderAttribute.<>c__DisplayClass6_0.<CreateActivator>b__4(Object target, IReadOnlyDependencyContainer dc)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.Activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyContainer.Inject[T](T instance)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Drawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Drawable.load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Drawable.Load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies, Boolean isDirectAsyncContext)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.loadChild(Drawable child)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.load(ShaderManager shaders, Nullable`1 cancellation)
2022-02-15 08:27:07 [error]: --- End of stack trace from previous location ---
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.BackgroundDependencyLoaderAttribute.<>c__DisplayClass6_0.<CreateActivator>b__4(Object target, IReadOnlyDependencyContainer dc)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.Activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyContainer.Inject[T](T instance)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Drawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Drawable.load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Drawable.Load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies, Boolean isDirectAsyncContext)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.loadChild(Drawable child)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.load(ShaderManager shaders, Nullable`1 cancellation)
2022-02-15 08:27:07 [error]: --- End of stack trace from previous location ---
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.BackgroundDependencyLoaderAttribute.<>c__DisplayClass6_0.<CreateActivator>b__4(Object target, IReadOnlyDependencyContainer dc)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.Activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyContainer.Inject[T](T instance)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Drawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Drawable.load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Drawable.Load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies, Boolean isDirectAsyncContext)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.loadChild(Drawable child)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.load(ShaderManager shaders, Nullable`1 cancellation)
2022-02-15 08:27:07 [error]: --- End of stack trace from previous location ---
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.BackgroundDependencyLoaderAttribute.<>c__DisplayClass6_0.<CreateActivator>b__4(Object target, IReadOnlyDependencyContainer dc)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyActivator.Activate(Object obj, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Allocation.DependencyContainer.Inject[T](T instance)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Drawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Drawable.load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies)
2022-02-15 08:27:07 [error]: at osu.Framework.Graphics.Drawable.Load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies, Boolean isDirectAsyncContext)
2022-02-15 08:27:07 [error]: at osu.Framework.Platform.GameHost.bootstrapSceneGraph(Game game)
2022-02-15 08:27:07 [error]: at osu.Framework.Platform.GameHost.Run(Game game)
2022-02-15 08:27:07 [error]: at osu.Game.Tests.VisualTestRunner.Main(String[] args) in D:\Dev\OSU\osu\osu.Game\Tests\VisualTestRunner.cs:line 17
```

**Comment by PercyDan54** [comment]
> Your IDE shows a detached hash that doesn't exist on this branch. Please check your git, it's probably in a bad state.

I'm not quite familiar with Git but I checked out at `2c3e50a`, the unknown hash is probably I used the merge HEAD
image

**Comment by peppy** [comment]
Will need to wait for someone else to test (that is known to have a working android build chain). It's hard to say if what you are seeing is due to the version of xamarin you are building against or not.

**Comment by Susko3** [comment]
> If anyone is able to test EF migration on android (after the recent commits), that would be exceedingly helpful.
> 
> Attached is a small database you can drop in your data folder if you don't have one any more. Make sure to also move away `client.realm` to allow the migration to actually run.
> 
> [client.db.zip](https://github.com/ppy/osu/files/8066397/client.db.zip)

Migration works fine on LG G7 Fit (API 28):

- uninstalled osu!
- checked out 2022.118.0 (last release before Realm)
- build and install Release configuration
- downloaded and played a map
- checked out this pull
- build and install Release
- test replay works

Also tested that the provided `client.db` migrated fine.

**Comment by peppy** [comment]
Perfect, thanks!

@smoogipoo should be good to go at your discretion.

**Comment by PercyDan54** [comment]
Can also confirm it works
Huawei P20(API 29)

File: Templates/Rulesets/ruleset-empty/osu.Game.Rulesets.EmptyFreeform.Tests/osu.Game.Rulesets.EmptyFreeform.Tests.csproj [hunk 1]
```diff
@@ -12,15 +12,15 @@
     <PackageReference Include="Appveyor.TestLogger" Version="2.0.0" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.0.0" />
     <PackageReference Include="NUnit" Version="3.13.2" />
-    <PackageReference Include="NUnit3TestAdapter" Version="4.1.0" />
+    <PackageReference Include="NUnit3TestAdapter" Version="4.2.1" />
```
Comment (smoogipoo) [inline Templates/Rulesets/ruleset-empty/osu.Game.Rulesets.EmptyFreeform.Tests/osu.Game.Rulesets.EmptyFreeform.Tests.csproj:15]: Needs revert

```suggestion
    <PackageReference Include="NUnit3TestAdapter" Version="4.1.0" />
```

File: Templates/Rulesets/ruleset-example/osu.Game.Rulesets.Pippidon.Tests/osu.Game.Rulesets.Pippidon.Tests.csproj [hunk 2]
```diff
@@ -12,15 +12,15 @@
     <PackageReference Include="Appveyor.TestLogger" Version="2.0.0" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.0.0" />
     <PackageReference Include="NUnit" Version="3.13.2" />
-    <PackageReference Include="NUnit3TestAdapter" Version="4.1.0" />
+    <PackageReference Include="NUnit3TestAdapter" Version="4.2.1" />
```
Comment (smoogipoo) [inline Templates/Rulesets/ruleset-example/osu.Game.Rulesets.Pippidon.Tests/osu.Game.Rulesets.Pippidon.Tests.csproj:15]: Needs revert

```suggestion
    <PackageReference Include="NUnit3TestAdapter" Version="4.1.0" />
```

File: Templates/Rulesets/ruleset-scrolling-empty/osu.Game.Rulesets.EmptyScrolling.Tests/osu.Game.Rulesets.EmptyScrolling.Tests.csproj [hunk 3]
```diff
@@ -12,15 +12,15 @@
     <PackageReference Include="Appveyor.TestLogger" Version="2.0.0" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.0.0" />
     <PackageReference Include="NUnit" Version="3.13.2" />
-    <PackageReference Include="NUnit3TestAdapter" Version="4.1.0" />
+    <PackageReference Include="NUnit3TestAdapter" Version="4.2.1" />
```
Comment (smoogipoo) [inline Templates/Rulesets/ruleset-scrolling-empty/osu.Game.Rulesets.EmptyScrolling.Tests/osu.Game.Rulesets.EmptyScrolling.Tests.csproj:15]: Needs revert

```suggestion
    <PackageReference Include="NUnit3TestAdapter" Version="4.1.0" />
```

File: Templates/Rulesets/ruleset-scrolling-example/osu.Game.Rulesets.Pippidon.Tests/osu.Game.Rulesets.Pippidon.Tests.csproj [hunk 4]
```diff
@@ -12,15 +12,15 @@
     <PackageReference Include="Appveyor.TestLogger" Version="2.0.0" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.0.0" />
     <PackageReference Include="NUnit" Version="3.13.2" />
-    <PackageReference Include="NUnit3TestAdapter" Version="4.1.0" />
+    <PackageReference Include="NUnit3TestAdapter" Version="4.2.1" />
```
Comment (smoogipoo) [inline Templates/Rulesets/ruleset-scrolling-example/osu.Game.Rulesets.Pippidon.Tests/osu.Game.Rulesets.Pippidon.Tests.csproj:15]: Needs revert

```suggestion
    <PackageReference Include="NUnit3TestAdapter" Version="4.1.0" />
```

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**